plan/expression_rewriter.go

```diff
@@ -3,32 +3,58 @@ package plan
 import (
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/evaluator"
 	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/util/types"
 )
 
-func (b *planBuilder) rewrite(expr ast.ExprNode, schema expression.Schema, AggrMapper map[*ast.AggregateFuncExpr]int) (newExpr expression.Expression, err error) {
-	er := &expressionRewriter{aggrMap: AggrMapper, schema: schema}
+// EvalSubquery evaluates incorrelated subqueries once.
+var EvalSubquery func(p Plan, is infoschema.InfoSchema, ctx context.Context) (types.Datum, error)
+
+func (b *planBuilder) rewrite(expr ast.ExprNode, p Plan, aggMapper map[*ast.AggregateFuncExpr]int) (newExpr expression.Expression, newPlan Plan, correlated bool, err error) {
+	er := &expressionRewriter{p: p, aggrMap: aggMapper, schema: p.GetSchema(), b: b}
 	expr.Accept(er)
 	if er.err != nil {
-		return nil, errors.Trace(er.err)
+		return nil, nil, false, errors.Trace(er.err)
 	}
 	if len(er.ctxStack) != 1 {
-		return nil, errors.Errorf("context len %v is invalid", len(er.ctxStack))
+		return nil, nil, false, errors.Errorf("context len %v is invalid", len(er.ctxStack))
 	}
-	return er.ctxStack[0], nil
+	return er.ctxStack[0], er.p, er.correlated, nil
 }
 
 type expressionRewriter struct {
-	ctxStack []expression.Expression
-	p        Plan
-	schema   expression.Schema
-	err      error
-	aggrMap  map[*ast.AggregateFuncExpr]int
-	isSubq   bool
-	b        *planBuilder
+	ctxStack   []expression.Expression
+	p          Plan
+	schema     expression.Schema
+	err        error
+	aggrMap    map[*ast.AggregateFuncExpr]int
+	b          *planBuilder
+	correlated bool
+}
+
+func (er *expressionRewriter) buildSubqueries(subq *ast.SubqueryExpr) (Plan, expression.Schema) {
+	if len(er.b.outerSchema) > 0 {
+		er.err = errors.New("Nested subqueries is not supported.")
+		return nil, nil
+	}
+	outerSchema := er.schema.DeepCopy()
+	for _, s := range outerSchema {
+		s.Correlated = true
+	}
+	nb := &planBuilder{id: er.b.id, outerSchema: outerSchema}
```

Comment:
put outerSchemas in a stack in planBuilder, so we don't need to create a new planbuilder to build subquery.